rspace/src/main/scala/coop/rchain/rspace/history/RadixTree.scala

```diff
@@ -0,0 +1,987 @@
+    *
+    * Must contain 256 [[Item]]s.
+    */
+  type Node = Vector[Item]
+
+  /**
+    * Number of items always constant.
+    */
+  val numItems = 256
+
+  /**
+    * Empty node consists only of [[EmptyItem]]s.
+    */
+  val emptyNode: Node = (0 until numItems).map(_ => EmptyItem).toVector
+
+  /**
+    * Binary codecs for serializing/deserializing Node in Radix tree
+    *
+    * {{{
+    * Coding structure for items:
+    *   EmptyItem                   - Empty (not encode)
+
+    *   Leaf(prefix,value)    -> [item number] [second byte] [prefix0]..[prefixM] [value0]..[value31]
+    *                               where is: [second byte] -> bit7 = 0 (Leaf identifier)
+    *                                                          bit6..bit0 - prefix length = M (from 0 to 127)
+    *
+    *   NodePtr(prefix,ptr)   -> [item number] [second byte] [prefix0]..[prefixM] [ptr0]..[ptr31]
+    *                               where is: [second byte] -> bit7 = 1 (NodePtr identifier)
+    *                                                          bit6..bit0 - prefix length = M (from 0 to 127)
+    *
+    * For example encode this Node which contains 2 non-empty items (number 1 and number 2):
+    * (0)[Empty] (1)[Leaf(prefix:0xFFFF,value:0x00..0001)] (2)[NodePtr(prefix:empty,value:0xFF..FFFF)] (3)...(255)[Empty].
+    * Encoded data = 0x0102FFFF0000000000000000000000000000000000000000000000000000000000000001
+    *                  0280FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
+    * where: item 1 (number_secondByte_prefix_value) = 01_02_FFFF_00..0001
+    *        item 2 (number_secondByte_prefix_value) = 02_80_empty_FF..FFFF
+    * }}}
+    */
+  object Codecs {
+    private val defSize  = 32 // Default size for non-empty item data
+    private val headSize = 2  // 2 bytes: first - item number, second - second byte
+
+    /** Serialization [[Node]] to [[ByteVector]]
+      */
+    def encode(node: Node): ByteVector = {
+
+      // Calculating size of serialized data
+      val calcSize = node.foldLeft(0) {
+        case (size, EmptyItem) => size
+
+        case (size, Leaf(leafPrefix, value)) =>
+          val sizePrefix = leafPrefix.size.toInt
+          assert(sizePrefix <= 127, "Error during serialization: size of prefix more than 127.")
+          assert(
+            value.size == defSize,
+            "Error during serialization: size of leafValue not equal 32."
+          )
+          size + headSize + sizePrefix + defSize
+
+        case (size, NodePtr(ptrPrefix, ptr)) =>
+          val sizePrefix = ptrPrefix.size.toInt
+          assert(sizePrefix <= 127, "Error during serialization: size of prefix more than 127.")
+          assert(
+            ptr.size == defSize,
+            "Error during serialization: size of ptrPrefix not equal 32."
+          )
+          size + headSize + sizePrefix + defSize
+      }
+
+      val arr = new Array[Byte](calcSize) //Allocation memory
```

Comment:
@tgrospic I think for some performance-critical places, we need to use mutable data to gain better performance.